File: frontend/src/services/inventory/equipmentService.ts
// frontend/src/services/inventory/equipmentService.ts

import { EquipmentService as EquipmentApiService } from './inventoryApi';
import { handleApiError } from '../api';
import type {
  EquipmentWithDetails,
  EquipmentCreateRequest,
  EquipmentUpdateRequest,
  EquipmentSearchParams,
  AssignmentRequest,
  FormatType
} from './types';

/**
 * Servicio de equipos con manejo de errores y transformación de datos
 */
export class EquipmentService {
  
  /**
   * Obtiene lista de equipos con formato específico
   */
  static async getList(params: {
    skip?: number;
    limit?: number;
    formatType?: FormatType;
  } = {}) {
    try {
      const { skip = 0, limit = 25, formatType = 'list' } = params;
      
      const equipment = await EquipmentApiService.getEquipmentList({
        skip,
        limit,
        format_type: formatType
      });
      
      return {
        success: true,
        data: equipment,
        total: equipment.length,
        message: `${equipment.length} equipos obtenidos exitosamente`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error('💥 Error en EquipmentService.getList:', errorMessage);
      
      return {
        success: false,
        data: [],
        total: 0,
        error: errorMessage
      };
    }
  }

  /**
   * Busca equipos con filtros avanzados
   */
  static async search(params: EquipmentSearchParams) {
    try {
      const results = await EquipmentApiService.searchEquipment(params);
      
      return {
        success: true,
        data: results.results,
        total: results.total_found,
        hasMore: results.has_more,
        nextCursor: results.next_cursor,
        message: `${results.total_found} equipos encontrados`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error('💥 Error en EquipmentService.search:', errorMessage);
      
      return {
        success: false,
        data: [],
        total: 0,
        hasMore: false,
        error: errorMessage
      };
    }
  }

  /**
   * Obtiene un equipo por ID
   */
  static async getById(id: number): Promise<{
    success: boolean;
    data?: EquipmentWithDetails;
    error?: string;
  }> {
    try {
      const equipment = await EquipmentApiService.getEquipmentById(id);
      
      return {
        success: true,
        data: equipment
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error(`💥 Error en EquipmentService.getById(${id}):`, errorMessage);
      
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * Crea un nuevo equipo
   */
  static async create(data: EquipmentCreateRequest): Promise<{
    success: boolean;
    data?: EquipmentWithDetails;
    error?: string;
    message?: string;
  }> {
    try {
      // Validaciones básicas
      if (!data.category_id) {
        throw new Error('La categoría es requerida');
      }
      if (!data.state_id) {
        throw new Error('El estado es requerido');
      }
      if (!data.location_id) {
        throw new Error('La ubicación es requerida');
      }

      const equipment = await EquipmentApiService.createEquipment(data);
      
      return {
        success: true,
        data: equipment,
        message: `Equipo ${equipment.codigo_ug || equipment.id} creado exitosamente`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error('💥 Error en EquipmentService.create:', errorMessage);
      
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * Actualiza un equipo existente
   */
  static async update(id: number, data: EquipmentUpdateRequest): Promise<{
    success: boolean;
    data?: EquipmentWithDetails;
    error?: string;
    message?: string;
  }> {
    try {
      const equipment = await EquipmentApiService.updateEquipment(id, data);
      
      return {
        success: true,
        data: equipment,
        message: `Equipo ${equipment.codigo_ug || equipment.id} actualizado exitosamente`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error(`💥 Error en EquipmentService.update(${id}):`, errorMessage);
      
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * Elimina un equipo
   */
  static async delete(id: number): Promise<{
    success: boolean;
    message?: string;
    error?: string;
  }> {
    try {
      const result = await EquipmentApiService.deleteEquipment(id);
      
      return {
        success: true,
        message: result.message || `Equipo ${id} eliminado exitosamente`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error(`💥 Error en EquipmentService.delete(${id}):`, errorMessage);
      
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * Asigna un equipo a un usuario
   */
  static async assign(equipmentId: number, userId: number, notes?: string): Promise<{
    success: boolean;
    data?: EquipmentWithDetails;
    error?: string;
    message?: string;
  }> {
    try {
      const assignmentData: AssignmentRequest = {
        user_id: userId,
        notes
      };

      const equipment = await EquipmentApiService.assignEquipment(equipmentId, assignmentData);
      
      return {
        success: true,
        data: equipment,
        message: `Equipo ${equipment.codigo_ug || equipment.id} asignado exitosamente`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error(`💥 Error en EquipmentService.assign(${equipmentId}, ${userId}):`, errorMessage);
      
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * Desasigna un equipo
   */
  static async unassign(equipmentId: number): Promise<{
    success: boolean;
    data?: EquipmentWithDetails;
    error?: string;
    message?: string;
  }> {
    try {
      const equipment = await EquipmentApiService.unassignEquipment(equipmentId);
      
      return {
        success: true,
        data: equipment,
        message: `Equipo ${equipment.codigo_ug || equipment.id} desasignado exitosamente`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error(`💥 Error en EquipmentService.unassign(${equipmentId}):`, errorMessage);
      
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * Obtiene equipos disponibles para asignación
   */
  static async getAvailable(limit: number = 50) {
    try {
      const results = await EquipmentApiService.searchEquipment({
        operational_only: true,
        unassigned_only: true,
        limit
      });
      
      return {
        success: true,
        data: results.results,
        total: results.total_found,
        message: `${results.total_found} equipos disponibles`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error('💥 Error en EquipmentService.getAvailable:', errorMessage);
      
      return {
        success: false,
        data: [],
        total: 0,
        error: errorMessage
      };
    }
  }

  /**
   * Obtiene equipos que necesitan mantenimiento
   */
  static async getNeedingMaintenance(limit: number = 50) {
    try {
      const results = await EquipmentApiService.searchEquipment({
        operational_only: false,
        limit
      });
      
      // Filtrar solo los no operativos (en la implementación real esto vendría del backend)
      const needingMaintenance = results.results.filter(eq => 
        !eq.state?.is_operational
      );
      
      return {
        success: true,
        data: needingMaintenance,
        total: needingMaintenance.length,
        message: `${needingMaintenance.length} equipos necesitan mantenimiento`
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error('💥 Error en EquipmentService.getNeedingMaintenance:', errorMessage);
      
      return {
        success: false,
        data: [],
        total: 0,
        error: errorMessage
      };
    }
  }

  /**
   * Operaciones en lote
   */
  static async bulkAssign(equipmentIds: number[], userId: number, notes?: string): Promise<{
    success: boolean;
    successfulCount: number;
    failedCount: number;
    errors: string[];
    message?: string;
  }> {
    const results = {
      success: true,
      successfulCount: 0,
      failedCount: 0,
      errors: [] as string[]
    };

    for (const equipmentId of equipmentIds) {
      try {
        await this.assign(equipmentId, userId, notes);
        results.successfulCount++;
      } catch (error) {
        results.failedCount++;
        results.errors.push(`Equipo ${equipmentId}: ${handleApiError(error)}`);
      }
    }

    if (results.failedCount > 0) {
      results.success = false;
    }

    return {
      ...results,
      message: `${results.successfulCount} equipos asignados, ${results.failedCount} fallos`
    };
  }

  /**
   * Obtiene estadísticas rápidas de equipos
   */
  static async getStats() {
    try {
      // En una implementación real, esto vendría de un endpoint específico
      const allEquipment = await this.getList({ limit: 1000 });
      
      if (!allEquipment.success) {
        throw new Error(allEquipment.error);
      }

      const equipment = allEquipment.data;
      const stats = {
        total: equipment.length,
        operational: equipment.filter((eq: any) => eq.state?.is_operational).length,
        assigned: equipment.filter((eq: any) => eq.assigned_user_id).length,
        available: equipment.filter((eq: any) => eq.state?.is_operational && !eq.assigned_user_id).length,
        needingMaintenance: equipment.filter((eq: any) => !eq.state?.is_operational).length
      };

      return {
        success: true,
        data: stats,
        message: 'Estadísticas obtenidas exitosamente'
      };
    } catch (error) {
      const errorMessage = handleApiError(error);
      console.error('💥 Error en EquipmentService.getStats:', errorMessage);
      
      return {
        success: false,
        data: {
          total: 0,
          operational: 0,
          assigned: 0,
          available: 0,
          needingMaintenance: 0
        },
        error: errorMessage
      };
    }
  }
}

// Export por defecto
export default EquipmentService;

File: backend/app/services/inventory/equipment_service.py
# ============================================================================
# backend/app/services/inventory/equipment_service.py
# ============================================================================

from typing import List, Optional, Dict, Any, Tuple
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from app.repositories.inventory.equipment_repository import EquipmentRepository
from app.models.inventory.equipment import Equipment
from app.utils.inventory_formatters import InventoryFormatter

class EquipmentService:
    """
    Servicio para lógica de negocio de equipos
    """
    
    @staticmethod
    def get_equipment_list(
        db: Session, 
        skip: int = 0, 
        limit: int = 25,
        format_type: str = "list"
    ) -> List[Dict[str, Any]]:
        """
        Obtiene lista de equipos formateada
        """
        equipment_list = EquipmentRepository.get_all_optimized(db, skip, limit)
        return InventoryFormatter.format_equipment_list(equipment_list, format_type)
    
    @staticmethod
    def get_equipment_by_id(db: Session, equipment_id: int) -> Dict[str, Any]:
        """
        Obtiene equipo por ID con detalles completos
        """
        equipment = EquipmentRepository.get_by_id_with_details(db, equipment_id)
        if not equipment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Equipo no encontrado"
            )
        
        return InventoryFormatter.format_equipment_details(equipment)
    
    @staticmethod
    def search_equipment(
        db: Session,
        search_query: str = None,
        filters: Dict[str, Any] = None,
        cursor: int = 0,
        limit: int = 25
    ) -> Dict[str, Any]:
        """
        Búsqueda de equipos con paginación cursor
        """
        if filters is None:
            filters = {}
        
        # Extraer filtros
        category_id = filters.get("category_id")
        state_id = filters.get("state_id")
        location_id = filters.get("location_id")
        assigned_user_id = filters.get("assigned_user_id")
        supplier_id = filters.get("supplier_id")
        
        # Realizar búsqueda
        equipment_list, total_count = EquipmentRepository.search_equipment(
            db=db,
            search_query=search_query,
            category_id=category_id,
            state_id=state_id,
            location_id=location_id,
            assigned_user_id=assigned_user_id,
            supplier_id=supplier_id,
            cursor=cursor,
            limit=limit
        )
        
        # Formatear resultados
        formatted_equipment = InventoryFormatter.format_equipment_list(equipment_list, "search")
        
        # Determinar si hay más resultados
        has_more = len(equipment_list) == limit
        next_cursor = equipment_list[-1].id if equipment_list and has_more else None
        
        return {
            "total_found": total_count,
            "results": formatted_equipment,
            "has_more": has_more,
            "next_cursor": next_cursor,
            "performance_hint": "consider_refining_search" if total_count > 500 else "showing_batch"
        }
    
    @staticmethod
    def create_equipment(
        db: Session, 
        equipment_data: Dict[str, Any], 
        created_by_id: int = None
    ) -> Dict[str, Any]:
        """
        Crea un nuevo equipo con validaciones
        """
        # Validaciones de negocio
        EquipmentService._validate_equipment_data(db, equipment_data)
        
        # Generar código UG si no se proporciona
        if not equipment_data.get("codigo_ug"):
            equipment_data["codigo_ug"] = EquipmentService._generate_equipment_code(
                db, equipment_data.get("category_id")
            )
        
        # Crear equipo
        equipment = EquipmentRepository.create(db, equipment_data, created_by_id)
        
        # Crear detalles de laboratorio si es necesario
        if equipment_data.get("lab_details"):
            EquipmentService._create_lab_details(db, equipment.id, equipment_data["lab_details"])
        
        return InventoryFormatter.format_equipment_details(equipment)
    
    @staticmethod
    def update_equipment(
        db: Session,
        equipment_id: int,
        equipment_data: Dict[str, Any],
        updated_by_id: int = None
    ) -> Dict[str, Any]:
        """
        Actualiza un equipo existente
        """
        equipment = EquipmentRepository.get_by_id_with_details(db, equipment_id)
        if not equipment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Equipo no encontrado"
            )
        
        # Validar datos actualizados
        EquipmentService._validate_equipment_data(db, equipment_data, equipment_id)
        
        # Actualizar equipo
        updated_equipment = EquipmentRepository.update(db, equipment, equipment_data, updated_by_id)
        
        return InventoryFormatter.format_equipment_details(updated_equipment)
    
    @staticmethod
    def delete_equipment(
        db: Session,
        equipment_id: int,
        deleted_by_id: int = None
    ) -> Dict[str, str]:
        """
        Elimina un equipo (soft delete)
        """
        equipment = EquipmentRepository.get_by_id_with_details(db, equipment_id)
        if not equipment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Equipo no encontrado"
            )
        
        # Verificar que no esté asignado
        if equipment.assigned_user_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No se puede eliminar un equipo asignado"
            )
        
        EquipmentRepository.soft_delete(db, equipment, deleted_by_id)
        
        return {
            "message": f"Equipo {equipment.codigo_ug} eliminado exitosamente"
        }
    
    @staticmethod
    def assign_equipment(
        db: Session,
        equipment_id: int,
        user_id: int,
        updated_by_id: int = None
    ) -> Dict[str, Any]:
        """
        Asigna un equipo a un usuario
        """
        equipment = EquipmentRepository.get_by_id_with_details(db, equipment_id)
        if not equipment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Equipo no encontrado"
            )
        
        # Verificar que el equipo esté disponible
        if equipment.assigned_user_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="El equipo ya está asignado"
            )
        
        # Verificar que el estado sea operativo
        if not equipment.state.is_operational:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No se puede asignar un equipo no operativo"
            )
        
        # Asignar equipo
        update_data = {"assigned_user_id": user_id}
        updated_equipment = EquipmentRepository.update(db, equipment, update_data, updated_by_id)
        
        return InventoryFormatter.format_equipment_details(updated_equipment)
    
    @staticmethod
    def unassign_equipment(
        db: Session,
        equipment_id: int,
        updated_by_id: int = None
    ) -> Dict[str, Any]:
        """
        Desasigna un equipo
        """
        equipment = EquipmentRepository.get_by_id_with_details(db, equipment_id)
        if not equipment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Equipo no encontrado"
            )
        
        if not equipment.assigned_user_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="El equipo no está asignado"
            )
        
        # Desasignar equipo
        update_data = {"assigned_user_id": None}
        updated_equipment = EquipmentRepository.update(db, equipment, update_data, updated_by_id)
        
        return InventoryFormatter.format_equipment_details(updated_equipment)
    
    # ===== MÉTODOS PRIVADOS =====
    
    @staticmethod
    def _validate_equipment_data(db: Session, data: Dict[str, Any], equipment_id: int = None):
        """
        Valida los datos de un equipo
        """
        # Validar código UG único
        if data.get("codigo_ug"):
            from app.repositories.inventory.equipment_repository import EquipmentRepository
            existing = db.query(Equipment).filter(
                Equipment.codigo_ug == data["codigo_ug"],
                Equipment.deleted_at.is_(None)
            ).first()
            
            if existing and (not equipment_id or existing.id != equipment_id):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="El código UG ya está en uso"
                )
        
        # Validar relaciones foráneas
        if data.get("category_id"):
            from app.repositories.inventory.category_repository import CategoryRepository
            if not CategoryRepository.get_by_id(db, data["category_id"]):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Categoría no válida"
                )
    
    @staticmethod
    def _generate_equipment_code(db: Session, category_id: int) -> str:
        """
        Genera un código UG automático
        """
        from app.repositories.inventory.category_repository import CategoryRepository
        
        category = CategoryRepository.get_by_id(db, category_id)
        prefix = category.name[:2].upper() if category else "EQ"
        
        # Obtener el último número
        last_equipment = db.query(Equipment).filter(
            Equipment.codigo_ug.like(f"{prefix}-%"),
            Equipment.deleted_at.is_(None)
        ).order_by(Equipment.id.desc()).first()
        
        if last_equipment and last_equipment.codigo_ug:
            try:
                last_number = int(last_equipment.codigo_ug.split('-')[-1])
                next_number = last_number + 1
            except (ValueError, IndexError):
                next_number = 1
        else:
            next_number = 1
        
        return f"{prefix}-{next_number:04d}"
    
    @staticmethod
    def _create_lab_details(db: Session, equipment_id: int, lab_data: Dict[str, Any]):
        """
        Crea detalles de laboratorio para un equipo
        """
        from app.models.inventory.equipment import EquipmentLab
        from datetime import datetime
        
        lab_data["equipment_id"] = equipment_id
        lab_data["created_at"] = datetime.utcnow()
        lab_data["updated_at"] = datetime.utcnow()
        
        db_lab = EquipmentLab(**lab_data)
        db.add(db_lab)
        db.commit()
        return db_lab